File: sdks/go/marketplace.go
package settlekit

import (
	"context"
	"net/http"
)

// CreateListingInput is the body for CreateListing (POST /v1/marketplace/listings).
type CreateListingInput struct {
	OrganizationID string         `json:"organizationId"`
	MerchantID     string         `json:"merchantId"`
	ProductID      string         `json:"productId"`
	Title          string         `json:"title"`
	Summary        string         `json:"summary,omitempty"`
	Description    string         `json:"description,omitempty"`
	Category       string         `json:"category,omitempty"`
	Tags           []string       `json:"tags,omitempty"`
	Metadata       map[string]any `json:"metadata,omitempty"`
}

// CreateListing creates a draft marketplace listing for a product.
// POST /v1/marketplace/listings.
func (c *Client) CreateListing(ctx context.Context, in CreateListingInput) (*Listing, error) {
	var out Listing
	if err := c.do(ctx, http.MethodPost, "/v1/marketplace/listings", in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ListListingsOptions filters and sorts ListListings. All fields are optional;
// Sort is one of "top", "new", or "price".
type ListListingsOptions struct {
	Query string
	Tag   string
	Sort  string
}

// ListListings lists published marketplace listings, optionally filtered by a
// full-text query and tag and sorted by "top", "new", or "price".
// GET /v1/marketplace/listings[?q=&tag=&sort=].
func (c *Client) ListListings(ctx context.Context, opts ListListingsOptions) ([]Listing, error) {
	q := newQuery()
	q.add("q", opts.Query)
	q.add("tag", opts.Tag)
	q.add("sort", opts.Sort)
	var out []Listing
	if err := c.do(ctx, http.MethodGet, "/v1/marketplace/listings"+q.encode(), nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// GetListing fetches a single listing by id.
// GET /v1/marketplace/listings/:id.
func (c *Client) GetListing(ctx context.Context, id string) (*Listing, error) {
	var out Listing
	if err := c.do(ctx, http.MethodGet, "/v1/marketplace/listings/"+id, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// PublishListing publishes a draft listing, making it publicly discoverable.
// POST /v1/marketplace/listings/:id/publish.
func (c *Client) PublishListing(ctx context.Context, id string) (*Listing, error) {
	var out Listing
	if err := c.do(ctx, http.MethodPost, "/v1/marketplace/listings/"+id+"/publish", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// RateListing submits a 1-5 star rating for a listing, updating its aggregate
// rating. POST /v1/marketplace/listings/:id/rate.
func (c *Client) RateListing(ctx context.Context, id string, stars int) (*Listing, error) {
	body := map[string]any{"stars": stars}
	var out Listing
	if err := c.do(ctx, http.MethodPost, "/v1/marketplace/listings/"+id+"/rate", body, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// SellerProfile fetches a merchant's public marketplace seller profile.
// GET /v1/marketplace/sellers/:merchantId.
func (c *Client) SellerProfile(ctx context.Context, merchantID string) (*SellerProfile, error) {
	var out SellerProfile
	if err := c.do(ctx, http.MethodGet, "/v1/marketplace/sellers/"+merchantID, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: sdks/python/settlekit/resources/marketplace.py
"""Marketplace resource — list, search, publish, and rate marketplace listings."""

from __future__ import annotations

from typing import Any, Optional, Sequence

from ._base import AsyncResource, SyncResource, clean_params


def _create_body(
    *,
    organization_id: str,
    merchant_id: str,
    product_id: str,
    title: str,
    summary: str,
    tags: Sequence[str],
) -> dict[str, Any]:
    return {
        "organizationId": organization_id,
        "merchantId": merchant_id,
        "productId": product_id,
        "title": title,
        "summary": summary,
        "tags": list(tags),
    }


class Marketplace(SyncResource):
    """Synchronous marketplace namespace."""

    def create_listing(
        self,
        *,
        organization_id: str,
        merchant_id: str,
        product_id: str,
        title: str,
        summary: str,
        tags: Sequence[str] = (),
    ) -> Any:
        """Create a listing. ``POST /v1/marketplace/listings``."""
        return self._request(
            "POST",
            "/v1/marketplace/listings",
            body=_create_body(
                organization_id=organization_id,
                merchant_id=merchant_id,
                product_id=product_id,
                title=title,
                summary=summary,
                tags=tags,
            ),
        )

    def list_listings(
        self,
        *,
        q: Optional[str] = None,
        tag: Optional[str] = None,
        sort: Optional[str] = None,
    ) -> Any:
        """Search listings. ``GET /v1/marketplace/listings?q=&tag=&sort=top|new|price``."""
        params = clean_params({"q": q, "tag": tag, "sort": sort})
        return self._request("GET", "/v1/marketplace/listings", params=params)

    def get_listing(self, listing_id: str) -> Any:
        """Fetch a listing by id. ``GET /v1/marketplace/listings/:id``."""
        return self._request("GET", f"/v1/marketplace/listings/{listing_id}")

    def publish_listing(self, listing_id: str) -> Any:
        """Publish a listing. ``POST /v1/marketplace/listings/:id/publish``."""
        return self._request(
            "POST", f"/v1/marketplace/listings/{listing_id}/publish"
        )

    def rate_listing(self, listing_id: str, *, stars: int) -> Any:
        """Rate a listing. ``POST /v1/marketplace/listings/:id/rate``."""
        return self._request(
            "POST",
            f"/v1/marketplace/listings/{listing_id}/rate",
            body={"stars": stars},
        )

    def get_seller(self, merchant_id: str) -> Any:
        """Fetch a seller profile. ``GET /v1/marketplace/sellers/:merchantId``."""
        return self._request("GET", f"/v1/marketplace/sellers/{merchant_id}")


class AsyncMarketplace(AsyncResource):
    """Asynchronous marketplace namespace."""

    async def create_listing(
        self,
        *,
        organization_id: str,
        merchant_id: str,
        product_id: str,
        title: str,
        summary: str,
        tags: Sequence[str] = (),
    ) -> Any:
        """Create a listing. ``POST /v1/marketplace/listings``."""
        return await self._request(
            "POST",
            "/v1/marketplace/listings",
            body=_create_body(
                organization_id=organization_id,
                merchant_id=merchant_id,
                product_id=product_id,
                title=title,
                summary=summary,
                tags=tags,
            ),
        )

    async def list_listings(
        self,
        *,
        q: Optional[str] = None,
        tag: Optional[str] = None,
        sort: Optional[str] = None,
    ) -> Any:
        """Search listings. ``GET /v1/marketplace/listings``."""
        params = clean_params({"q": q, "tag": tag, "sort": sort})
        return await self._request(
            "GET", "/v1/marketplace/listings", params=params
        )

    async def get_listing(self, listing_id: str) -> Any:
        """Fetch a listing by id. ``GET /v1/marketplace/listings/:id``."""
        return await self._request(
            "GET", f"/v1/marketplace/listings/{listing_id}"
        )

    async def publish_listing(self, listing_id: str) -> Any:
        """Publish a listing. ``POST /v1/marketplace/listings/:id/publish``."""
        return await self._request(
            "POST", f"/v1/marketplace/listings/{listing_id}/publish"
        )

    async def rate_listing(self, listing_id: str, *, stars: int) -> Any:
        """Rate a listing. ``POST /v1/marketplace/listings/:id/rate``."""
        return await self._request(
            "POST",
            f"/v1/marketplace/listings/{listing_id}/rate",
            body={"stars": stars},
        )

    async def get_seller(self, merchant_id: str) -> Any:
        """Fetch a seller profile. ``GET /v1/marketplace/sellers/:merchantId``."""
        return await self._request(
            "GET", f"/v1/marketplace/sellers/{merchant_id}"
        )
